design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+# MLB-0001: BGP FRR enablement
+
+## Summary
+
+The purpose of this enhancement is to use [Free Range
+Routing](https://frrouting.org/) (FRR) as an alternative BGP implementation
+in MetalLB. When directed to, MetalLB will publish prefixes via FRR rather than
+MetalLB’s current built-in BGP speaker.
+
+## Motivation
+
+The motivation for this enhancement is to provide an alternative
+production-ready BGP agent for use in MetalLB. Overall, this should reduce the
+maintenance cost for the MetalLB project by reducing the effort required for
+adding additional BGP features and bug fixes. For example, there are a number of
+issues in the current backlog that may be addressed by using FRR:
+
+* [Add support for MP BGP encoding for IPv4 and IPv6](https://github.com/metallb/metallb/pull/590)
+* [BFD support](https://github.com/metallb/metallb/issues/396)
+* [BGP Failover too slow](https://github.com/metallb/metallb/issues/485)
+* [OSPF Support](https://github.com/metallb/metallb/issues/256)
+
+FRR is a mature Linux Foundation routing protocol suite based on Quagga that has
+been used in many production deployments. As such, it has been proven in terms
+of its maturity, flexibility (as can be seen by the [broad range of features](http://docs.frrouting.org/en/latest/bgp.html)
+it supports), scalability, security, reliability and performance. It also
+provides detailed logging features to aid debugging.
+
+The current native BGP implementation of MetalLB has limitations when coexisting
+with other BGP agents on the node. If these BGP agents are attempting to peer
+with external routers, this may invoke BGP’s conflict resolution algorithm. This
+is detailed [here](https://metallb.universe.tf/configuration/calico/) for the
+Calico case and [here](https://metallb.universe.tf/configuration/kube-router/)
+for the Kube-Router case. This enhancement may be a step towards resolving these
```

Comment:
In what way does the FRR integration address e.g. the MetalLB + Calico problem? Not sure I got that. Could you elaborate?

Reply:
I think I see your confusion and I will update. From my understanding of the problem, only one agent can peer with the remote. Therefore, if MetalLB's native implementation is pairing with the remote, no other BGP software can establish a session. As the native implementation is currently limited and is tightly coupled with MetalLB so it could not be shared with another entity wishing to advertise with the remote. This enhancement improves this situation because FRR could potentially be shared by another entity allowing other prefixes to be advertised via this session. In the Calico case, this would require plugging in their own (BIRD?) BGP implementation as an alternative MetalLB BGP implementation which could then be reused by Calico.

Reply:
Sharing the same FRR instance may create other challenges like detecting conflicting configurations and also properly tracking the advertised routes owner (to make sure a component does not delete routes from an other component, and to properly detect staling routes)

Reply:
Yes, I agree but I am just specifying here that this is a "step towards" resolving these issues as it should provide a way to eventually solve this. I will add this clarification. Is that OK?